scrapy/core/spidermw.py

```diff
@@ -3,15 +3,40 @@
 
 See documentation in docs/topics/spider-middleware.rst
 """
+from itertools import chain
+
 import six
 from twisted.python.failure import Failure
+from scrapy.exceptions import _InvalidOutput
 from scrapy.middleware import MiddlewareManager
 from scrapy.utils.defer import mustbe_deferred
 from scrapy.utils.conf import build_component_list
 
+
 def _isiterable(possible_iterator):
     return hasattr(possible_iterator, '__iter__')
 
+
+class MutableChain:
+    """
+    Thin wrapper around itertools.chain, allowing to add iterables "in-place"
+    """
+    def __init__(self, *args):
+        self.data = chain(*args)
+
+    def extend(self, *iterables):
+        self.data = chain(self.data, *iterables)
+
+    def __iter__(self):
+        return self.data.__iter__()
+
+    def __next__(self):  # py3
+        return self.data.__next__()
+
+    def next(self):  # py2
```

Comment:
I wonder if you can make `def next` an alias `next = __next__`, and implement `__next__` as `next(self.data)`. `next` built-in function calls `.next()` in Python 2 and `__next__()` in Python 3 - see e.g. https://docs.python.org/2/library/functions.html#next